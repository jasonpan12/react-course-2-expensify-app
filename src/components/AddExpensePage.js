import React from "react";
import {connect} from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import {startAddExpense} from "../actions/expenses";

export class AddExpensePage extends React.Component {
	onSubmit = (expense) => {
		this.props.startAddExpense(expense);
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<h1>Add Expense</h1>
				<ExpenseForm
					onSubmit={this.onSubmit}
				/>
			</div>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startAddExpense: (expense) => dispatch(startAddExpense(expense)) // implicitly return an object, where onSubmit is a function
	// pull addExpense into here so that we can test this component
	// should be named addExpense, the name of the dispatcher
	// return your dispatcher functions here
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
