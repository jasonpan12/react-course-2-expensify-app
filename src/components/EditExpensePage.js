import React from "react";
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {startEditExpense, startRemoveExpense} from "../actions/expenses";

// Refactor EditExpensePage to be a class based component, which will let us pull out the inline fnctions
// set up mapDispatchToProps editExpense and removeExpense

// should render EditExpensePage
// snapshot

// should handle editExpense
// spies

// should handle removeExpense
// spies

export class EditExpensePage extends React.Component {
	onSubmit = (expense) => { // by binding the onSubmit here, we aren't re-rendering onsubmit each time the props update
		// the expense object is provided down in ExpenseForm
		// dispatch action to edit expense
		// redirect to dashboard page
		this.props.startEditExpense(this.props.expense.id, expense)
		this.props.history.push('/');
	}

	onRemove = () => {
		// dispatch action to removeExpense
		this.props.startRemoveExpense({id: this.props.expense.id})
		this.props.history.push('/');
	}

	render() {
		return (
			<div>
				<div className = "page-header">
					<div className = "content-container">
						<h1 className = "page-header__title">Edit Expense</h1>
					</div>
				</div>
				<div className = "content-container">
					<ExpenseForm
						expense = {this.props.expense}
						onSubmit = {this.onSubmit}
					/>
					<div>
						<button className = "button button--secondary" onClick={this.onRemove}>Remove Expense</button>
					</div>
				</div>
			</div>
		)
	}
}

const mapStateToProps = (state, props) => { // these props can be sent down to the HOC
	return { // EditExpensePage is given an expense prop
		expense: state.expenses.find((expense) => expense.id === props.match.params.id)
		// All expenses are in the state, but this page has an id in its params
		// the expense prop is whichever expense matches the ID in the params.
	}
}

const mapDispatchToProps = (dispatch, props) => ({
	startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
	startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage); // this is the HOC (EditExpensePage), and we can pass new props in
