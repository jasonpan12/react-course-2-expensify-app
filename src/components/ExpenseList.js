import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from '../selectors/expenses';

// Export this so we can test the unconnected version in jest
export const ExpenseList = (props) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			) : (
				props.expenses.map((expense) => {
					return <ExpenseListItem key={expense.id} {...expense}/> // spread operator
				})
			)
		}
	</div>
);

// Takes state and makes it into props for the connect api
// basically what do you want off the store, then what component do you want to connect the store to
const mapStateToProps = (state) => ({
	// Determine which expenses we should show in the list
	// by using selectExpenses
	expenses: selectExpenses(state.expenses, state.filters)
})

// call it ConnectedExpenseList
// connect returns a function, which means we need to call that function
export default connect(mapStateToProps)(ExpenseList);
