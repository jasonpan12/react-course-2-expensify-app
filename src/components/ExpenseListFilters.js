import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
	state = {
		calendarFocused: null // ExpenseForm  also kept track of calendarFocused
	}
	onDatesChange = ({startDate, endDate}) => { // onDatesChange will be called by the react-dates library with an object startDate and endDate
		this.props.setStartDate(startDate);
		this.props.setEndDate(endDate);
	}
	onFocusChange = (calendarFocused) => {
		this.setState(() => ({calendarFocused}))
	}

	onTextChange = (event) => {
		// dispatch gets included in props with connected components
		this.props.setTextFilter(event.target.value);
	}

	onSortChange = (event) => {
		if (event.target.value === "date") {
			this.props.sortByDate()
		} else if (event.target.value === "amount") {
			this.props.sortByAmount()
		}
	}

	render() {
		return (
			<div>
				<input placeholder="Text Filter" type="text" value={this.props.filters.text}
					   onChange = {this.onTextChange}/>
				<select value={this.props.filters.sortBy} onChange = {this.onSortChange}>
					<option value="date">Date</option>
					<option value="amount">Amount</option>
				</select>
				<DateRangePicker
					startDate={this.props.filters.startDate}
					endDate={this.props.filters.endDate}
					onDatesChange={this.onDatesChange}
					focusedInput={this.state.calendarFocused}
					onFocusChange={this.onFocusChange}
					showClearDates={true}
					numberOfMonths={1}
					isOutsideRange={() => false}
				/>
			</div>
		);
	}
}

// Pulls state from redux store, passes in filters as props
const mapStateToProps = (state) => {
	return {
		filters: state.filters
	};
}

const mapDispatchToProps = (dispatch) => ({ // factor all dispatch calls down here
	setTextFilter: (text) => dispatch(setTextFilter(text)),
	sortByDate: () => dispatch(sortByDate()),
	sortByAmount: () => dispatch(sortByAmount()),
	setStartDate: (startDate) =>  dispatch(setStartDate(startDate)),
	setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
// Map dispatch functions to Props
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
