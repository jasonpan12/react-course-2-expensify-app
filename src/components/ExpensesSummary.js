import React from 'react';
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from "../selectors/expenses-total";
import {connect} from "react-redux";
import numeral from "numeral";

// Must export this on its own (unconnected) in order to test the unconnected component
export const ExpensesSummary = (props) => {
	const expenseCount = props.expenses.length;
	const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
	const expensesTotal = numeral(selectExpensesTotal(props.expenses)/100).format('$0,0.00');
	return (
		<h1>{`Viewing ${expenseCount} expenses totalling ${expensesTotal}`}</h1>
	);
}

// Map redux state to props. props.expenses is the result of selectExpenses.
const mapStateToProps = (state) => ({
	expenses: selectExpenses(state.expenses, state.filters)
})

// Connect this component to the Redux store
export default connect(mapStateToProps)(ExpensesSummary);
