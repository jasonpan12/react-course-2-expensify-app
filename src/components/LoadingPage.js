import React from 'react';

const LoadingPage = () => (
	<div className = "loader">
		<p><img className = "loader__image" src="/images/loader.gif"/></p>
	</div>
)
export default LoadingPage;
