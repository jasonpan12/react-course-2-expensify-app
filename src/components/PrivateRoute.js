import React from 'react';
import {connect} from 'react-redux';
import {Route, Redirect} from "react-router-dom";
import Header from "../components/Header";

// This is how you rename a lower case prop to upper case Component
// component comes from approuter.js
// isAuthenticated comes from mapStatetoProps
// We want to pull out isAuthenticated and component, and then get the rest of the props
export const PrivateRoute = ({
	isAuthenticated,
	component: Component,
	...rest // Rest operator
}) => (
	<Route {...rest} component={(props) => (
		isAuthenticated ? (
			<div>
				<Header/>
				<Component {...props} />
			</div>
		) : (
			<Redirect to="/"/>
		)
		)}/> // Pass in rest of props
);

const mapStateToProps = (state) => ({
	isAuthenticated: !!state.auth.uid
});

export default connect(mapStateToProps)(PrivateRoute);
