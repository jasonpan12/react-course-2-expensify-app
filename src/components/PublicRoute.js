import React from 'react';
import {connect} from 'react-redux';
import {Route, Redirect} from "react-router-dom";
import Header from "../components/Header";

// Route to Dashboard if authenticated
// Route to requested component if not authenticated
export const PublicRoute = ({
	isAuthenticated,
	component: Component,
	...rest // Rest operator
}) => (
	<Route {...rest} component={(props) => (
		isAuthenticated ? (
			<Redirect to="/dashboard"/>
		) : (
			<div>
				<Header/>
				<Component {...props} />
			</div>
		)
		)}/> // Pass in rest of props
);

const mapStateToProps = (state) => ({
	isAuthenticated: !!state.auth.uid
});

export default connect(mapStateToProps)(PublicRoute);

// Create PublicRoute
// Redirect to /dashboard
// Render component if not logged in
// Use it for the Login page
