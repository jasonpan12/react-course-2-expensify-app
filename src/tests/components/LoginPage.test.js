import {shallow} from "enzyme/build";
import React from "react";
import {LoginPage} from "../../components/LoginPage";

let wrapper, startLogin;

beforeEach(() => {
	startLogin = jest.fn();

	wrapper = shallow(<LoginPage startLogin={startLogin}/>);
})
test('should render login page that matches snapshot', () => {
	expect(wrapper).toMatchSnapshot();
});

test('should call startLogin on button click', () => {
	wrapper.find('button').simulate('click');
	expect(startLogin).toHaveBeenCalled();
})
