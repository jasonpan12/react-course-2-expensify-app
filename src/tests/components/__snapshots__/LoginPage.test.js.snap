// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render login page that matches snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
